### ISSUE TITLE
feat: Volume, Speech Rate, and Pitch Controls for Text-to-Speech (TTS) Output

### ISSUE THREAD

**Issue by silentoplayz** [open-webui/open-webui#1331]
**Problem Description:**
The current version of Open WebUI lacks the necessary customization options for the text-to-speech (TTS) output, including volume control, speech rate adjustment, pitch adjustment, and audio playback functionality for speaking out notifications. These limitations hinder the user experience and accessibility of the text-to-speech (TTS) feature.

**Describe the solution you'd like:**
I propose the implementation of the following features to enhance the TTS output customization:
1. A volume control slider to adjust the volume of the TTS output.
2. A "Speech Rate" slider to adjust the speed of the TTS output.
3. A "Pitch" slider enabling users to modify the voice pitch of the TTS output.
4. An option to enable or disable audio playback for speaking out notifications.

**Alternative solution:**
Offer predefined volume, speed, & pitch options instead of a slider for a simpler interface.

**Alternatives Considered:**
Manually adjusting the device's overall volume or utilizing third-party applications to manipulate speech output and volume settings represents a workaround. However, this solution proves to be inconvenient for users, necessitating the addition of these much-needed features within Open WebUI.

**Additional Context:**
This feature request focuses on improving the text-to-speech (TTS) feature's accessibility and overall user experience. Implementing these requested features, including volume, speed, and pitch adjustments, will significantly enhance user satisfaction and convenience. It's crucial to maintain compatibility with existing features, ensuring this customization suite does not adversely impact any existing functionalities or behaviors.

**Comment by dannyl1u** [comment]
I think this is a feature, how does this look for the UI?

<img width="729" alt="image" src="https://github.com/open-webui/open-webui/assets/45186464/0545b604-8bc7-4b3b-a911-eb09c0da98f3">

**Comment by dannyl1u** [comment]
I think this is would be a good feature, how does this look for the UI?

<img width="703" alt="image" src="https://github.com/open-webui/open-webui/assets/45186464/91ffdb4a-8581-4e40-80a3-a797a2ec4ae8">

**Comment by Silentoplayz** [comment]
That looks good to me @dannyl1u, although, do you think the sliders could take on a similar form as the model advanced parameter sliders?

![Screenshot 2024-03-16 141517](https://github.com/open-webui/open-webui/assets/50341825/35cb12ff-b9c0-4ac6-89a8-6acde682a866)

**Comment by dannyl1u** [comment]
> That looks good to me @dannyl1u, although, do you think the sliders could take on a similar form as the model advanced parameter sliders? I only ask because I feel that tjbck would step in to ask the same thing eventually or even make the adjustment themselves.
> 
> ![Screenshot 2024-03-16 141517](https://private-user-images.githubusercontent.com/50341825/320372810-35cb12ff-b9c0-4ac6-89a8-6acde682a866.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI1NTkwNTMsIm5iZiI6MTcxMjU1ODc1MywicGF0aCI6Ii81MDM0MTgyNS8zMjAzNzI4MTAtMzVjYjEyZmYtYjljMC00YWM2LTg5YTgtNmFjZGU2ODJhODY2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDA4VDA2NDU1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTEzYjljMGU5YmNiNzZkY2UxM2I3M2I2MTdlZWQwMmY3NTUzYTFhNmYxNmJhMjY4OWY5MjJmYTg0NzQ0NzM0ZjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Ilh_540Rz13kYE0vhyhpTYloyVsEQXXJhaRrp-Xygs8)
> 
> P.S: Thank you for your contributions to Open WebUI!

Yes! Thanks for the suggestion, I forgot those sliders existed 😆 , that's definitely the better UI and I'll reuse that!

**Comment by UXVirtual** [comment]
@dannyl1u another challenge with TTS output I've noticed is generated markdown code blocks are spoken out audibly.

Making this a toggle option, and stripping the code block prior to the `extractSentences()` call if it is toggled on would help with coding assistant use-cases.

**Comment by littledot2020** [comment]
> @dannyl1u我注意到的 TTS 输出的另一个挑战是生成的 markdown 代码块是以声音形式读出的。
> 
> 将其设为切换选项，并在`extractSentences()`切换后剥离调用之前的代码块，这将有助于编码助手用例。
I also want to know how to play content formatted after converting markdown.

**Comment by thiswillbeyourgithub** [comment]
I definitely think that there needs to be a button that if you click it makes those options appear in a call mode. This way you can adjust the settings without having to exit the call mode. Currently the call mode is fairly unusable in my opinion because of the lack of customizability on the fly. Also, there definitely needs to be an option to display the output as it's being read and received from the LLM. In many situations, the text-to-speech can fail just a few words and having to exit the call mode just to read the missing snippet is really an issue.

**Comment by silentoplayz** [comment]
Related - https://github.com/open-webui/open-webui/pull/5509 (speech playback speed control for Call mode)

**Comment by Cdddo** [comment]
The TTS speed and volume should be accessible from the chat view, which is where it is needed most.

**Comment by Winor** [comment]
> The TTS speed and volume should be accessible from the chat view, which is where it is needed most.

I agree, we should have a tts media player.

**Comment by kyunwang** [comment]
> * A "Speech Rate" slider to adjust the speed of the TTS output.

Started a PR to allow for these two controls in chat view including volume control specifically. #8512 

![Screenshot 2025-01-12 at 21 01 45](https://github.com/user-attachments/assets/202fb396-6e43-44f6-9762-68b684382bcc)
![Screenshot 2025-01-12 at 23 17 00](https://github.com/user-attachments/assets/7e0ed1bf-3000-4c88-972c-d3a308434bab)

**Comment by stefancrain** [comment]
In the hopes that this helps others finding this issue in the short term, [this comment](https://github.com/remsky/Kokoro-FastAPI/issues/75#issuecomment-2605260554) enabled me to configure the TTS speedup I needed.  
 
> ... each user can configure the playback speed of TTS in their own settings menu (Click profile image > Settings > Audio > Speech Playback Speed). ...

I can work on a PR to add the user based playback speed instructions to the docs for [Kokoro-FastAPI-integration](https://docs.openwebui.com/tutorials/text-to-speech/Kokoro-FastAPI-integration) (my use-case). As this change is not specific to that TTS provider, is there a better home for that guide?

**Comment by emory** [comment]
I don't know if I could get some support for something related to this or if I should create a different enhancement request. It's a different interface for a user to adjust and alter the resulting audio output of a TTS service using a prompt.

I really hate to use this for an example, but it's the best implementation of what I'm curious about implementing or enabling. In the customization options for a *Rabbit R1* when the user is logged into *Rabbithole* [^hole], you can prompt the user interface elements of responses to queries ("make it look like a tricorder"), and you can also *prompt the voice*.

I have a favorite, so here's what I do: in the Rabbit config prompt inbox box, I put a prompt. Type of affect, speech characteristics or linguistic traits, but you can use colloquialisms too you don't need to be a speech pathologist especially for basic effects processing operations like changes to the speed and/or pitch of the output. The result of this prompting is actually really interesting. I have one I use that makes `Shimmer` sound, in my opinion, a dead ringer for Dr. Sharon Fieldstone on Apple tv+'s Ted Lasso. I usually use masculine voices so my kids get used to bossing men around, but I struck gold. (hint: the first thing I prompt is "with a posh accent," which it interprets as, I believe `Transatlantic` and/or `Upper RP` but like I said, colloquialisms work too.)

So obviously this could also be used to define whatever attributes you want that diverge from the typical: "17% faster tempo, clear enunciation, use aliteration where you can and have fun with it, you're from new orleans and it's mardis gras!"

…rather than having to dial/slide things in?

Am I in the wrong place? I don't want to open unnecessary issues but I might be too late to this party.

//emory

[hole]: a portal/playground environment where you do irresponsible things in an environment of dubious integrity that you have to assume is completely public.
